File: benchmarks/compute/nbody/Python/nbody.py
import time
import random
import math
import sys

# ---------- 可选数据规模 ----------
DATASET = 'STANDARD'

if DATASET == 'SMALL':
    BODIES = 100
    ITERATIONS = 10
elif DATASET == 'LARGE':
    BODIES = 5000
    ITERATIONS = 100
else:
    BODIES = 1000
    ITERATIONS = 50

class Vec3:
    def __init__(self, x=0.0, y=0.0, z=0.0):
        self.x = x
        self.y = y
        self.z = z

class Body:
    def __init__(self):
        self.position = Vec3()
        self.velocity = Vec3()
        self.acceleration = Vec3()
        self.mass = 0.0

# ---------- 初始化天体 ----------
def init_bodies(n, seed):
    random.seed(seed)
    bodies = [Body() for _ in range(n)]
    for i in range(n):
        bodies[i].position.x = random.random() * 100.0
        bodies[i].position.y = random.random() * 100.0
        bodies[i].position.z = random.random() * 100.0
        
        bodies[i].velocity.x = random.random() * 10.0
        bodies[i].velocity.y = random.random() * 10.0
        bodies[i].velocity.z = random.random() * 10.0
        
        bodies[i].acceleration.x = 0.0
        bodies[i].acceleration.y = 0.0
        bodies[i].acceleration.z = 0.0
        
        bodies[i].mass = random.random() * 1000.0 + 100.0
    return bodies

# ---------- 打印部分输出 ----------
def print_result(bodies):
    for i in range(min(3, len(bodies))):
        print(f"Body {i}: pos=({bodies[i].position.x:.2f}, {bodies[i].position.y:.2f}, {bodies[i].position.z:.2f}) "
              f"vel=({bodies[i].velocity.x:.2f}, {bodies[i].velocity.y:.2f}, {bodies[i].velocity.z:.2f})")

# ---------- 主计算过程 ----------
def compute_nbody(n, iterations, bodies):
    G = 6.67430e-11  # 万有引力常数
    
    for _ in range(iterations):
        # 重置加速度
        for i in range(n):
            bodies[i].acceleration.x = 0.0
            bodies[i].acceleration.y = 0.0
            bodies[i].acceleration.z = 0.0
        
        # 计算引力
        for i in range(n):
            for j in range(i + 1, n):
                dx = bodies[j].position.x - bodies[i].position.x
                dy = bodies[j].position.y - bodies[i].position.y
                dz = bodies[j].position.z - bodies[i].position.z
                
                dist_sq = dx*dx + dy*dy + dz*dz + 1e-10  # 避免除以零
                dist = math.sqrt(dist_sq)
                force = G * bodies[i].mass * bodies[j].mass / dist_sq
                
                fx = force * dx / dist
                fy = force * dy / dist
                fz = force * dz / dist
                
                bodies[i].acceleration.x += fx / bodies[i].mass
                bodies[i].acceleration.y += fy / bodies[i].mass
                bodies[i].acceleration.z += fz / bodies[i].mass
                
                bodies[j].acceleration.x -= fx / bodies[j].mass
                bodies[j].acceleration.y -= fy / bodies[j].mass
                bodies[j].acceleration.z -= fz / bodies[j].mass
        
        # 更新速度和位置
        for i in range(n):
            bodies[i].velocity.x += bodies[i].acceleration.x
            bodies[i].velocity.y += bodies[i].acceleration.y
            bodies[i].velocity.z += bodies[i].acceleration.z
            
            bodies[i].position.x += bodies[i].velocity.x
            bodies[i].position.y += bodies[i].velocity.y
            bodies[i].position.z += bodies[i].velocity.z

# ---------- 主函数 ----------
if __name__ == "__main__":
    seed = int(sys.argv[1]) if len(sys.argv) > 1 else 42
    bodies = init_bodies(BODIES, seed)

    start = time.time()
    compute_nbody(BODIES, ITERATIONS, bodies)
    end = time.time()

    print(f"Finished N-body simulation in {end - start:.3f} seconds")
    print_result(bodies)

File: benchmarks/compute/nbody/Zig/nbody.zig
const std = @import("std");

// ---------- 可选数据规模 ----------
const SMALL = 0;
const STANDARD = 1;
const LARGE = 2;

const DATASET = STANDARD;

const BODIES = switch (DATASET) {
    SMALL => 100,
    LARGE => 5000,
    else => 1000,
};

const ITERATIONS = switch (DATASET) {
    SMALL => 10,
    LARGE => 100,
    else => 50,
};

// ---------- 向量和天体结构 ----------
const Vec3 = struct {
    x: f64,
    y: f64,
    z: f64,
};

const Body = struct {
    position: Vec3,
    velocity: Vec3,
    acceleration: Vec3,
    mass: f64,
};

// ---------- 初始化天体 ----------
fn initBodies(n: usize, bodies: *[BODIES]Body, seed: u32) void {
    var rng = std.rand.DefaultPrng.init(seed);
    for (0..n) |i| {
        bodies[i] = Body{
            .position = Vec3{
                .x = rng.random().float(f64) * 100.0,
                .y = rng.random().float(f64) * 100.0,
                .z = rng.random().float(f64) * 100.0,
            },
            .velocity = Vec3{
                .x = rng.random().float(f64) * 10.0,
                .y = rng.random().float(f64) * 10.0,
                .z = rng.random().float(f64) * 10.0,
            },
            .acceleration = Vec3{
                .x = 0.0,
                .y = 0.0,
                .z = 0.0,
            },
            .mass = rng.random().float(f64) * 1000.0 + 100.0,
        };
    }
}

// ---------- 打印部分输出 ----------
fn printResult(n: usize, bodies: *[BODIES]Body) void {
    const limit = @min(3, n);
    for (0..limit) |i| {
        std.debug.print("Body {d}: pos=({d:.2}, {d:.2}, {d:.2}) vel=({d:.2}, {d:.2}, {d:.2})\n", .{
            i,
            bodies[i].position.x,
            bodies[i].position.y,
            bodies[i].position.z,
            bodies[i].velocity.x,
            bodies[i].velocity.y,
            bodies[i].velocity.z,
        });
    }
}

// ---------- 主计算过程 ----------
fn computeNbody(n: usize, iterations: usize, bodies: *[BODIES]Body) void {
    const G = 6.67430e-11; // 万有引力常数

    for (0..iterations) |_| {
        // 重置加速度
        for (0..n) |i| {
            bodies[i].acceleration = Vec3{ .x = 0.0, .y = 0.0, .z = 0.0 };
        }

        // 计算引力
        for (0..n) |i| {
            for (i + 1..n) |j| {
                const dx = bodies[j].position.x - bodies[i].position.x;
                const dy = bodies[j].position.y - bodies[i].position.y;
                const dz = bodies[j].position.z - bodies[i].position.z;

                const dist_sq = dx * dx + dy * dy + dz * dz + 1e-10; // 避免除以零
                const dist = @sqrt(dist_sq);
                const force = G * bodies[i].mass * bodies[j].mass / dist_sq;

                const fx = force * dx / dist;
                const fy = force * dy / dist;
                const fz = force * dz / dist;

                bodies[i].acceleration.x += fx / bodies[i].mass;
                bodies[i].acceleration.y += fy / bodies[i].mass;
                bodies[i].acceleration.z += fz / bodies[i].mass;

                bodies[j].acceleration.x -= fx / bodies[j].mass;
                bodies[j].acceleration.y -= fy / bodies[j].mass;
                bodies[j].acceleration.z -= fz / bodies[j].mass;
            }
        }

        // 更新速度和位置
        for (0..n) |i| {
            bodies[i].velocity.x += bodies[i].acceleration.x;
            bodies[i].velocity.y += bodies[i].acceleration.y;
            bodies[i].velocity.z += bodies[i].acceleration.z;

            bodies[i].position.x += bodies[i].velocity.x;
            bodies[i].position.y += bodies[i].velocity.y;
            bodies[i].position.z += bodies[i].velocity.z;
        }
    }
}

// ---------- 主函数 ----------
pub fn main() !void {
    var bodies: [BODIES]Body = undefined;

    // WASI兼容的命令行参数解析
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const args = try std.process.argsAlloc(allocator);
    const seed = if (args.len > 1) 
        try std.fmt.parseInt(u32, args[1], 10)
    else 
        42;

    initBodies(BODIES, &bodies, seed);

    const start = std.time.nanoTimestamp();
    computeNbody(BODIES, ITERATIONS, &bodies);
    const end = std.time.nanoTimestamp();

    const elapsed_seconds = @as(f64, @floatFromInt(end - start)) / 1e9;
    std.debug.print("Finished N-body simulation in {d:.3} seconds\n", .{elapsed_seconds});

    printResult(BODIES, &bodies);
}
